x/metadata/spec/examples/kotlin/src/main/kotlin/MetadataAddress.kt

```diff
@@ -0,0 +1,209 @@
+        private fun uuidAsByteArray(uuid: UUID): ByteArray {
+            val b = ByteBuffer.wrap(ByteArray(16))
+            b.putLong(uuid.mostSignificantBits)
+            b.putLong(uuid.leastSignificantBits)
+            return b.array()
+        }
+
+        /** Converts a ByteArray to a UUID. */
+        private fun byteArrayAsUuid(data: ByteArray): UUID {
+            val uuidBytes = ByteArray(16)
+            if (data.size >= 16) {
+                data.copyInto(uuidBytes, 0, 0, 16)
+            } else if (data.isNotEmpty()) {
+                data.copyInto(uuidBytes, 0, 0, data.size)
+            }
+            val bb = ByteBuffer.wrap(uuidBytes)
+            val mostSig = bb.long
+            val leastSig = bb.long
+            return UUID(mostSig, leastSig)
+        }
+
+        /** Hashes a string and gets the bytes desired for a MetadataAddress. */
+        private fun asHashedBytes(str: String) =
+            MessageDigest.getInstance("SHA-256").digest(str.trim().toLowerCase().toByteArray()).copyOfRange(0, 16)
+
+        /** Get the prefix that corresponds to the provided key Byte. */
+        private fun getPrefixFromKey(key: Byte) =
+            when (key) {
+                KEY_SCOPE -> PREFIX_SCOPE
+                KEY_SESSION -> PREFIX_SESSION
+                KEY_RECORD -> PREFIX_RECORD
+                KEY_SCOPE_SPECIFICATION -> PREFIX_SCOPE_SPECIFICATION
+                KEY_CONTRACT_SPECIFICATION -> PREFIX_CONTRACT_SPECIFICATION
+                KEY_RECORD_SPECIFICATION -> PREFIX_RECORD_SPECIFICATION
+                else -> {
+                    throw IllegalArgumentException("Invalid key: $key")
+                }
+            }
+
+        /**
+         * Converts from a base64 encoded ByteArray to base32 encoded ByteArray and then to bech32 address string.
+         */
+        private fun convertAndEncode(humanReadablePart: String, dataBase64: ByteArray) =
+            Bech32.encode(humanReadablePart, convertBits(dataBase64, 8, 5, true))
+
+        /**
+         * Decodes a bech32 encoded string and converts to base64 encoded bytes.
+         * Note: Even though this returns an object named "Bech32Data" the data is actual base64 encoded.
+         */
+        private fun decodeAndConvert(bech32: String): Bech32Data {
+            val (hrp, dataBase32) = Bech32.decode(bech32)
+            val dataBase64 = convertBits(dataBase32, 5, 8, false)
+            return Bech32Data(hrp, dataBase64)
+        }
+
+        // convertBits Taken from [Bitcoinj SegwitAddress Java implementation](https://github.com/bitcoinj/bitcoinj/blob/31c7e5fbceb9884cb02d2dabc755009caa2d613e/core/src/main/java/org/bitcoinj/core/SegwitAddress.java)
+        /**
+         * converts a ByteArray where each byte is encoding fromBits bits,
+         * to a ByteArray where each byte is encoding toBits bits.
+         */
+        private fun convertBits(data: ByteArray, fromBits: Int, toBits: Int, pad: Boolean): ByteArray {
```

Comment:
This should be in the bech32 class as the 5bit conversion is an implementation detail of that encoding format.

There is another version of this bech32 code based on a reference implementation used throughout Figure (but referenced here)
https://github.com/provenance-io/explorer-service/blob/8d89dfff1e6ea4dc890c7b35b9ffaa5e712d88f3/service/src/main/kotlin/io/provenance/explorer/domain/core/Bech32.kt#L126-L179

Reply:
That's what I thought too, but the libraries I was basing this off (KBech32 and Bitcoinj) didn't have it in with the Bech32 stuff. KBech32 didn't have it at all, and Bitcoinj has it inside some SegwitAddress stuff (inaccessible from outside there). And since the classes in the KBech32 folder were copied, I didn't want to add a large chunk of new functionality to them.

That being said, I'll check out the class created and used in the explorer. I'll see about copying that one into here and using it instead of this copied KBech32 stuff.

Reply:
I deleted the KBech32 stuff and copied the Bech32 class out of explorer. I felt it was still a bit clunky, though, so I cleaned it up a bit.

1. I replaced the convertBits code with the one I used to have in the MetadataAddress class which let me remove all the `@ExperimentalUnsignedTypes` decorations.
2. I changed things around so that externally, no one needs to know about the 5-bit thing. The Bech32Data now has the 8-bit data and the encode method takes in 8-bit data.
3. Removed the `String` and `ByteArray` extensions. They just felt weird syntactically.
4. Removed the checksum from the Bech32Data since it doesn't make sense (to me) to expose that.
5. Got rid of the Provenance-specific values, e.g. `PROVENANCE_MAINNET_PREFIX`.
6. Made several functions private that don't need to be exposed (and might cause confusion if exposed).

I also cleaned up some stuff in the MetadataAddress.kt file that is now handled in the Bech32 stuff.